### PR TITLE
optimize trytype loop

### PR DESCRIPTION
**Pull request by a-zakir**
The current dispatch method iterates through a list of functions **trytype**, sequentially checking each one to see if it can handle the given node type. This can be inefficient.
in this PR we use unordered_map as look up table between node type and trytype function:
less comparisons && faster access than sequential iteration

### CODE REVIEW

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: There are two possibilities for errors
- The map does not contain an entry
- The function was found, but returns `std::nullopt`

We should have a different error message for each case. In the first case, you must catch `std::out_of_range` exception thrown by `std::map::at`.

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]:
```suggestion
        ([&my_map, &NodeTypes] { my_map[typeid(NodeTypes)] = &tryVisit<R, NodeVisitor<R, Args...>, NodeTypes>; }(),
```

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]:
```suggestion
        const static auto nodeDispatchFunctions = NodesVisitList<AddNode,
```

**Review by flomnes — COMMENTED** [review]
If you use `typeid` to determine the polymorphic type of node objects, you no longer need to check the result of `dynamic_cast` in templated function `tryVisit`.

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: i wrapped the the _if_ cond in a try/catch blocks.
if the  _if_ cond body is not computed an exception is raised
in the catch block, a message is printed and the original exception (which could originate from any part of the code) is re-thrown

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: Please use shorter blocks with more specific exception types
```cpp
try
{
    my_map.at(...)
}
catch (const std::out_of_range& ex) 
{
// ...
}
```

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: This could be a free function (outside `NodeVisitor`)

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h:113]:
```suggestion
            logs.error() << "Antares::Solver::Visitor: could not visit the node! (exception " << ex.what() << ")";
```

https://en.cppreference.com/w/cpp/error/exception/what

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: You may move this out of class `NodeVisitor`

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: two variadic parameters, it will look like
template<class R, class... NodeTypes, class... Args> // not ok

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h:113]: not necessary, when the exception is re-thrown;

**Review by flomnes — APPROVED** [review]

**Comment by flomnes** [comment]
Waiting for @guilpier-code 's approval

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (guilpier-code) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h:113]: About : 
```C++
logs.error() << "Antares::Solver::Visitor: could not visit the node!";
```
Could be interesting to know which node type is missing in the map. We don't have this information in the message.

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (guilpier-code) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: Very clever, I mean it.
But the code here became quite complex (it was a pain to understand the mechanisms at work in this PR).

A justification for all this complexity is an expected better efficiency.
We just talked about this with @flomnes : we wonder if, given the limited size of the map, we would get a performance benefit from that new change.
Maybe we could measure it.

**Review by guilpier-code — CHANGES_REQUESTED** [review]

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (guilpier-code) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: **NodeDispatchFunctionsProvider** could be renamed **NodeVisitsProvider** more in related the underlying design pattern.

Same for **nodeDispatchFunctions** : ==> **nodeVisitList** ?

**Review by guilpier-code — CHANGES_REQUESTED** [review]

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h:113]: yes, this will be done in [2354](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2354)

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h]: the number of node types is not large (nb_Node_type<20), but in a deep/large tree the benefit is to avoid nb_Node_type * nb_nodes [trytype loop ]

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (guilpier-code) [inline src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h:44]: **Question** :
To avoid too much complexity (essentially due to template, and the very powerful **variadic pack** C++ feature), could we not more simply have :
```C++
R dispatch(const Node& node, Args... args)
{
    static const std::unordered_map<std::type_index, FunctionT> nodeDispatchFunctions = 
    {
        {typeid(AddNode), &tryVisit<R, NodeVisitor<R, Args...>, AddNode>},
        {typeid(SubtractionNode), &tryVisit<R, NodeVisitor<R, Args...>, SubtractionNode>},
        {typeid(MultiplicationNode), &tryVisit<R, NodeVisitor<R, Args...>, MultiplicationNode>},
        ....
    }
    
    try
    {
        return nodeDispatchFunctions.at(typeid(node))(node, *this, args...);
    }
    catch (std::exception&)
    {
        logs.error() << "Antares::Solver::Visitor: could not visit the node!";
        throw;
    }
}
```
?
Would it even be possible ?

I suppose that the template **struct**, its template **NodesVisitList** function and the **variadic pack** system inside were meant to avoid **code duplication** in the **node type** that we can see in the code exerpt just above, and I can see that the goal is reached, but the price is a bit high in code complexity.

**Review by guilpier-code — CHANGES_REQUESTED** [review]

File: src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h:44]: yes that's exactly what is done by the compiler in NodeVisitsProvider::NodesVisitList

**Pull request merged**
